### PR TITLE
Pull media player out of the QuestionWidget

### LINKED ISSUE #2521: Bugs related to audio files
#### Software and hardware versions 
Collect v1.16.4 and older versions

#### Problem description
Playing with audio files I discovered a few bugs I'll try to describe.

**BUG 1:**
If you have a select_one (or select_multiple) question and each item has an audio file you have to click two times to start an audio if another item is playing.

Expected behavior:
A new audio file should start immediately and if another one is playing that another one should be stopped.

**BUG 2:**
If we have AudioWidget and that widget contains audioButton it doesn't work at all and very strange things happen that are even difficult to describe. For example, if you click on audioButton and then you want to play a recorded sound the recorder plays a file that comes from the audioButton.

**BUG 3:**
If we have a field-list with a few questions it's possible to play many audios at the same time.

Here is a form which allows reproducing all mentioned issues: 
[MediaForm.zip](https://github.com/opendatakit/collect/files/2428602/MediaForm.zip)

### LINKED ISSUE #2861: Autoplay option doesn't work
#### Software and hardware versions 
Collect v1.19.0 and older versions

#### Problem description
During testing #2817 we discovered that `autoplay` option doesn't work.
It crashes on 1.19, 1.18, 1.17
There is no crash but it also doesn't work on 1.16, 1.15 ... 1.8 (we didn't check every version between but we can assume it doesn't work in all cases)

#### Other information 
@yanokwa I said that probably we can remove that option since it seems useless but you didn't agree.
I think I'm right... @mmarciniak90 was testing that option today and it doesn't work on all devices in the way I described above (sometimes crashes).

It crashes on three last versions but we have no crash reports. Today @mmarciniak90 caused firs appearances:
https://console.firebase.google.com/u/0/project/api-project-322300403941/crashlytics/app/android:org.odk.collect.android/issues/5c5d46dbf8b88c29631622fc?time=last-seven-days&sessionId=5C5D2E4102EA00013CCDBAA626940E27_DNE_8_v2

https://console.firebase.google.com/u/0/project/api-project-322300403941/crashlytics/app/android:org.odk.collect.android/issues/5c5d47acf8b88c2963163597?time=last-seven-days

additionally, no one complained about it (it doesn't work on 1.16, 1.15 ... 1.8 and probably older versions) so I'm pretty sure it's completely useless. 

The link you attached https://forum.opendatakit.org/search?q=autoplay%20order%3Alatest also shows only one question about that option during last three years.

Even if we want to fix that option for some reason keep it then it would require some discussion.
1. Should that play only media files attached to choices or the one attached to the question too?
2. What if we have multiple questions on one page and we use that option.
etc.

I'm not a fan of that option.

### PR DESCRIPTION
**Pull request by shobhitagarwal1612**
Closes #2861 
Closes #2521

#### What has been done to verify that this works as intended?
Tested using 4 forms on Android 7, Redmi Note 4

Tests performed:
* Autoplay question audio/video if enabled
* Double pressing audio button stops it
* Swiping next while the audio is running stops the audio
* Text color changes to tint color (blue) when playing audio
* Minimizing the app while the audio is running stops the audio
* Select Widget / Grid Widget / Grid Multi Widget and Audio Widgets are functioning properly

#### Why is this the best possible solution? Were any other approaches considered?
Since a number of classes are using media player but only once instance should run at a time. So, moving the media player into a global module and injecting it wherever it is required. The listeners use RxEventBus to listen to various media events asynchronously

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?
This might affect the loading of audio widgets

#### Do we need any specific form for testing your changes? If so, please attach one.
I tested `All-Widgets-Form`, `Birds` form and the following ones
https://drive.google.com/file/d/14Pcu8hdE7gBqmFXpbPi6KW5779InmiWP/view?usp=sharing
https://github.com/opendatakit/collect/files/2850939/form.zip

#### Does this change require updates to documentation? If so, please file an issue [here]( https://github.com/opendatakit/docs/issues/new) and include the link below.
No

#### Before submitting this PR, please make sure you have:
- [x] run `./gradlew checkAll` and confirmed all checks still pass OR confirm CircleCI build passes and run `./gradlew connectedDebugAndroidTest` locally.
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/opendatakit/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/opendatakit/collect/blob/master/CONTRIBUTING.md#ui-components-style-guidelines)

### CODE REVIEW

**Comment by shobhitagarwal1612** [comment]
@grzesiek2010 As discussed on slack, I have moved media player out of the question widget instead creating another level of abstraction. Please review

**Comment by shobhitagarwal1612** [comment]
@grzesiek2010 The PR is ready for you to review.

I have tried to simplify the code by reducing a lot of duplicate calls.

**Comment by shobhitagarwal1612** [comment]
Autoplay tested using the form provided by @grzesiek2010 https://drive.google.com/file/d/14Pcu8hdE7gBqmFXpbPi6KW5779InmiWP/view?usp=sharing
https://github.com/opendatakit/collect/files/2850939/form.zip

**Comment by lognaturel** [comment]
@shobhitagarwal1612 can you please expand on what alternatives you considered to `RxEventBus` and how this addresses the concerns described at https://github.com/opendatakit/collect/issues/2521#issuecomment-429436203?

**Comment by shobhitagarwal1612** [comment]
> alternatives you considered to RxEventBus

The only alternative for `RxEventBus` I could think was to maintain a list of listeners in `MediaController` and add/remove the listeners from `QuestionWidget` manually on view creation/destruction. I chose to go with `RxEventBus` since it is already being used in the project and adding another design pattern (composite manual subscription) seemed unusual to me.

> it should be possible to leverage the existing relationship between the components involved to address the problem without it.

One of the problems with the existing components was that media controlling and widgets (including `ODKView`) were deeply coupled and it was really getting really complex to identify which of the set listener `MediaPlayer.OnCompletionListener, MediaPlayer.OnErrorListener, MediaPlayer.OnPreparedListener` was going to execute as it was overwritten in many places. Hence, having a separate module which exclusively handles the `MediaPlayer` made it easy to understand/read the code. This module `MediaController` broadcasts media events (completed, prepared) which are observed by the view (widgets) to update UI events or autoplay.

To summarize:
 - `MediaController` is injectable
 - `FormEntryActivity` can directly stop/release current media based on activity's state (paused/finished) instead of doing it for every widget
 - `MediaController` exposes methods to play/pause media
 - `MediaController` exposes a method to check if a file is being played or not.
 - Events are observed by widgets(view) using `RxJava`
 - `AudioHandler` was a redundant wrapper class as only the URI was unique in it and same copy of `MediaPlayer` was passed to each. Moreover, Handler was a confusing name as developers might try to compare it with android's [handler](https://developer.android.com/reference/android/os/Handler)
 -  Fixes bugs in autoplay and improves readbility

**Comment by lognaturel** [comment]
Thanks once again for taking on this ambitious refactor, @shobhitagarwal1612, and for your patience with slow feedback. This turns out to be a really great case study of coupling and how to detangle it.

Now that Android's architecture components are out of beta and that there has been more usage of them around the code base, I think it makes sense to favor `LiveData` for reactive patterns. While raw RxJava might still have its place eventually for more complex cases, hopefully we can move away from using the event bus implementation.

@seadowg has also honed in on this as an interesting puzzle and an opportunity to establish patterns that can be used for future detangling. His approach at #3229 leverages architecture components and some tests with the goal of fully covering all of the intended functionality.

In light of that, I am closing this PR. Thanks for laying out some great inspiration for how to slice responsibilities up and for identifying the media functionality as in need of a rethink. I think you'll recognize quite a few similarities in #3229 and if you have a moment, please do engage there.

**Comment by shobhitagarwal1612** [comment]
Sorry for the slow response. I wasn't very active since past few weeks. But I'm happy to see that @seadowg is already working on this issue 😄
